File: runtime/src/test/java/org/apache/flink/agents/runtime/operator/ReActSubagentTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.agents.runtime.operator;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.flink.agents.api.Event;
import org.apache.flink.agents.api.InputEvent;
import org.apache.flink.agents.api.OutputEvent;
import org.apache.flink.agents.api.agents.Agent;
import org.apache.flink.agents.api.agents.ReActAgent;
import org.apache.flink.agents.api.chat.messages.ChatMessage;
import org.apache.flink.agents.api.chat.messages.MessageRole;
import org.apache.flink.agents.api.chat.model.BaseChatModelConnection;
import org.apache.flink.agents.api.chat.model.BaseChatModelSetup;
import org.apache.flink.agents.api.context.RunnerContext;
import org.apache.flink.agents.api.resource.ResourceContext;
import org.apache.flink.agents.api.resource.ResourceDescriptor;
import org.apache.flink.agents.api.resource.ResourceType;
import org.apache.flink.agents.api.subagent.SubagentResult;
import org.apache.flink.agents.api.subagent.SubagentSetup;
import org.apache.flink.agents.api.tools.Tool;
import org.apache.flink.agents.plan.AgentPlan;
import org.apache.flink.agents.plan.actions.Action;
import org.apache.flink.agents.runtime.actionstate.ActionState;
import org.apache.flink.agents.runtime.actionstate.ActionStateSerde;
import org.apache.flink.agents.runtime.actionstate.InMemoryActionStateStore;
import org.apache.flink.agents.runtime.subagent.InternalSubagentCallEvent;
import org.apache.flink.agents.runtime.subagent.InternalSubagentSetup;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.api.java.functions.KeySelector;
import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
import org.apache.flink.streaming.util.KeyedOneInputStreamOperatorTestHarness;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assumptions.assumeTrue;

/** Exercises the actual scoped ReAct chat/tool loop without an external model. */
@Timeout(30)
public class ReActSubagentTest {
    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final List<String> CALLS = new ArrayList<>();
    private static boolean structured;

    public static class Answer {
        public String answer;
    }

    @BeforeEach
    void reset() {
        assumeTrue(Runtime.version().feature() >= 21, "Internal Java calls need continuations");
        CALLS.clear();
        structured = false;
    }

    public static class ScriptedSetup extends BaseChatModelSetup {
        public ScriptedSetup(ResourceDescriptor descriptor, ResourceContext context) {
            super(descriptor, context);
        }

        @Override
        public Map<String, Object> getParameters() {
            return new HashMap<>(Map.of("model", getModel()));
        }
    }

    public static class ScriptedConnection extends BaseChatModelConnection {
        public ScriptedConnection(ResourceDescriptor descriptor, ResourceContext context) {
            super(descriptor, context);
        }

        @Override
        public ChatMessage chat(
                List<ChatMessage> messages, List<Tool> tools, Map<String, Object> params) {
            String model = (String) params.get("model");
            ChatMessage last = messages.get(messages.size() - 1);
            CALLS.add(model + ":" + last.getRole().getValue());
            if (last.getRole() == MessageRole.TOOL) {
                assertThat(last.getContent())
                        .contains(model.equals("child") ? "child evidence" : "child answer");
                return new ChatMessage(
                        MessageRole.ASSISTANT,
                        structured && model.equals("child")
                                ? "{\"answer\":\"child answer\"}"
                                : model + " answer");
            }
            if (model.equals("child")) {
                assertThat(messages.get(0).getContent()).isEqualTo("Literal {prompt} instructions");
                assertThat(last.getContent()).isEqualTo("investigate");
                assertThat(tools).extracting(Tool::getName).containsExactly("evidence");
                return request("evidence", Map.of());
            }
            assertThat(tools).extracting(Tool::getName).containsExactly("_subagent_researcher");
            assertThat(tools.get(0).getDescription()).contains("Research a task");
            return request("_subagent_researcher", Map.of("prompt", "investigate"));
        }
    }

    public static class ChildTools {
        @org.apache.flink.agents.api.annotation.Tool(description = "Child evidence")
        public static String evidence() {
            return "child evidence";
        }
    }

    public static class ParentTools {
        @org.apache.flink.agents.api.annotation.Tool(description = "Parent evidence")
        public static String evidence() {
            return "parent evidence";
        }
    }

    private static ChatMessage request(String name, Map<String, Object> arguments) {
        return ChatMessage.assistant(
                "",
                List.of(
                        Map.of(
                                "id",
                                "call-1",
                                "type",
                                "function",
                                "function",
                                Map.of("name", name, "arguments", arguments))));
    }

    private static ResourceDescriptor model(String name) {
        ResourceDescriptor.Builder builder =
                ResourceDescriptor.Builder.newBuilder(ScriptedSetup.class.getName())
                        .addInitialArgument("connection", "connection")
                        .addInitialArgument("model", name);
        if (name.equals("child")) {
            builder.addInitialArgument("tools", List.of("evidence"));
        } else {
            builder.addInitialArgument("subagents", List.of("researcher"));
        }
        return builder.build();
    }

    private static Agent child() throws Exception {
        Agent child =
                ReActAgent.forSubagent(
                        model("child"),
                        "Research a task",
                        "Literal {prompt} instructions",
                        structured ? Answer.class : null);
        child.addAction(
                new String[] {InputEvent.EVENT_TYPE},
                ReActSubagentTest.class.getMethod(
                        "assertChildScope", Event.class, RunnerContext.class));
        child.addResource(
                "evidence",
                ResourceType.TOOL,
                Tool.fromMethod(ChildTools.class.getMethod("evidence")));
        return child;
    }

    private static void resources(Agent parent) throws Exception {
        parent.addResource(
                "connection",
                ResourceType.CHAT_MODEL_CONNECTION,
                new ResourceDescriptor(ScriptedConnection.class.getName(), Map.of()));
        parent.addResource(
                "evidence",
                ResourceType.TOOL,
                Tool.fromMethod(ParentTools.class.getMethod("evidence")));
        parent.addResource("researcher", ResourceType.AGENT, child());
    }

    public static void assertChildScope(Event event, RunnerContext ctx) throws Exception {
        assertThat(ctx.getSensoryMemory().isExist("parent_secret")).isFalse();
        assertThat(ctx.getSensoryMemory().isExist("child_marker")).isFalse();
        ctx.getSensoryMemory().set("child_marker", "private");
    }

    public static void invokeTwice(Event event, RunnerContext ctx) throws Exception {
        invoke(event, ctx);
        invoke(event, ctx);
    }

    public static void invoke(Event event, RunnerContext ctx) throws Exception {
        SubagentSetup child = (SubagentSetup) ctx.getResource("researcher", ResourceType.AGENT);
        ctx.getSensoryMemory().set("parent_secret", "private");
        Object input = InputEvent.fromEvent(event).getInput();
        SubagentResult result =
                child.submit(
                                ctx,
                                input.equals(2L)
                                        ? Map.of("prompt", 42)
                                        : Map.of("prompt", "investigate"))
                        .await();
        assertThat(ctx.getSensoryMemory().isExist("child_marker")).isFalse();
        assertThat(ctx.getSensoryMemory().isExist("_TOOL_CALL_CONTEXT")).isFalse();
        ctx.sendEvent(
                new OutputEvent(
                        result.isSuccess() ? result.getResult() : result.getErrorMessage()));
    }

    private static Agent explicitParent() throws Exception {
        Agent parent = new Agent();
        parent.addAction(
                new String[] {InputEvent.EVENT_TYPE},
                ReActSubagentTest.class.getMethod("invoke", Event.class, RunnerContext.class));
        resources(parent);
        return parent;
    }

    @Test
    void explicitCallRunsTheChildToolLoopAfterPlanReconstruction() throws Exception {
        assertThat(run(new AgentPlan(explicitParent()), 1L))
                .containsExactly(List.of("child answer"));
        assertThat(CALLS).containsExactly("child:user", "child:tool");
    }

    @Test
    void modelDelegationResumesTheParentWithTheChildAnswer() throws Exception {
        Agent parent = new ReActAgent(model("parent"), null, null);
        resources(parent);
        assertThat(run(new AgentPlan(parent), 1L)).containsExactly("parent answer");
        assertThat(CALLS).containsExactly("parent:user", "child:user", "child:tool", "parent:tool");
    }

    @Test
    void invalidPromptIsAChildFailureAndDoesNotCallTheModel() throws Exception {
        List<Object> output = run(new AgentPlan(explicitParent()), 2L);
        assertThat(output).hasSize(1);
        assertThat(output.get(0).toString()).contains("only a string 'prompt'");
        assertThat(CALLS).isEmpty();
    }

    @Test
    void successiveCallsHaveIndependentChildMemory() throws Exception {
        Agent parent = new Agent();
        parent.addAction(
                new String[] {InputEvent.EVENT_TYPE},
                ReActSubagentTest.class.getMethod("invokeTwice", Event.class, RunnerContext.class));
        resources(parent);
        assertThat(run(new AgentPlan(parent), 1L))
                .containsExactly(List.of("child answer"), List.of("child answer"));
        assertThat(CALLS).containsExactly("child:user", "child:tool", "child:user", "child:tool");
    }

    @Test
    void structuredChildOutputIsNormalizedForModelDelegation() throws Exception {
        structured = true;
        Agent parent = new ReActAgent(model("parent"), null, null);
        resources(parent);
        assertThat(run(new AgentPlan(parent), 1L)).containsExactly("parent answer");
    }

    /** Persist values by copy, as a durable backend does, before later actions mutate memory. */
    private static class SnapshotStore extends InMemoryActionStateStore {
        SnapshotStore() {
            super(false);
        }

        @Override
        public void put(Object key, long seq, Action action, Event event, ActionState state)
                throws IOException {
            super.put(
                    key,
                    seq,
                    action,
                    event,
                    ActionStateSerde.deserialize(ActionStateSerde.serialize(state)));
        }
    }

    @Test
    void replayRestoresChildMemoryBeforeResumingItsToolLoop() throws Exception {
        InMemoryActionStateStore first = new SnapshotStore();
        assertThat(run(new AgentPlan(explicitParent()), 1L, first))
                .containsExactly(List.of("child answer"));
        assertThat(CALLS).containsExactly("child:user", "child:tool");
        Map<String, ActionState> recoveredStates = new LinkedHashMap<>();
        first.getKeyedActionStates()
                .get(1L)
                .forEach(
                        (key, state) -> {
                            if (state.getTaskEvent() instanceof InternalSubagentCallEvent) {
                                InternalSubagentCallEvent envelope =
                                        (InternalSubagentCallEvent) state.getTaskEvent();
                                // Retain the initial child request's completed actions, but drop
                                // the tool call
                                // and later actions. The restored tool response needs the earlier
                                // memory writes.
                                if (envelope.getDelegateEventType().equals(InputEvent.EVENT_TYPE)
                                        || envelope.getDelegateEventType()
                                                .equals("_chat_request_event")) {
                                    recoveredStates.put(
                                            key,
                                            ActionStateSerde.deserialize(
                                                    ActionStateSerde.serialize(state)));
                                }
                            }
                        });
        assertThat(recoveredStates).hasSize(3);
        InMemoryActionStateStore recovered = new InMemoryActionStateStore(false);
        recovered.getKeyedActionStates().put(1L, recoveredStates);
        assertThat(run(new AgentPlan(explicitParent()), 1L, recovered))
                .containsExactly(List.of("child answer"));
        assertThat(CALLS).containsExactly("child:user", "child:tool", "child:tool");
    }

    @Test
    void failedChildRemainsFailedWhenOnlyItsStateWasPersisted() throws Exception {
        InMemoryActionStateStore first = new SnapshotStore();
        List<Object> failure = run(new AgentPlan(explicitParent()), 2L, first);
        assertThat(failure.get(0).toString()).contains("only a string 'prompt'");
        Map<String, ActionState> childStates = new LinkedHashMap<>();
        first.getKeyedActionStates()
                .get(2L)
                .forEach(
                        (key, state) -> {
                            if (state.getTaskEvent() instanceof InternalSubagentCallEvent) {
                                childStates.put(
                                        key,
                                        ActionStateSerde.deserialize(
                                                ActionStateSerde.serialize(state)));
                            }
                        });
        assertThat(childStates.values())
                .anyMatch(state -> state.getSubagentFailureMessage() != null);
        InMemoryActionStateStore recovered = new InMemoryActionStateStore(false);
        recovered.getKeyedActionStates().put(2L, childStates);
        assertThat(run(new AgentPlan(explicitParent()), 2L, recovered)).isEqualTo(failure);
        assertThat(CALLS).isEmpty();
    }

    @Test
    void descriptorRoundTripPreservesCallableMetadataAndChildPlan() throws Exception {
        AgentPlan plan = new AgentPlan(explicitParent());
        InternalSubagentSetup setup =
                (InternalSubagentSetup)
                        plan.getResourceProviders()
                                .get(ResourceType.AGENT)
                                .get("researcher")
                                .provide(null);
        ResourceDescriptor descriptor =
                MAPPER.readValue(
                        MAPPER.writeValueAsString(setup.getDescriptor()), ResourceDescriptor.class);
        InternalSubagentSetup restored = new InternalSubagentSetup(descriptor, null);
        assertThat(restored.getDescription()).isEqualTo("Research a task");
        assertThat(MAPPER.readTree(restored.getInputSchema()).path("required").get(0).asText())
                .isEqualTo("prompt");
        assertThat(restored.getChildPlan().getActions()).containsKey("startAction");
    }

    @SuppressWarnings("unchecked")
    private static List<Object> run(AgentPlan plan, long input) throws Exception {
        return run(plan, input, new InMemoryActionStateStore(false));
    }

    @SuppressWarnings("unchecked")
    private static List<Object> run(AgentPlan plan, long input, InMemoryActionStateStore store)
            throws Exception {
        AgentPlan restored = MAPPER.readValue(MAPPER.writeValueAsString(plan), AgentPlan.class);
        try (KeyedOneInputStreamOperatorTestHarness<Long, Long, Object> harness =
                new KeyedOneInputStreamOperatorTestHarness<>(
                        new ActionExecutionOperatorFactory<>(restored, true, store),
                        (KeySelector<Long, Long>) value -> value,
                        TypeInformation.of(Long.class))) {
            harness.open();
            harness.processElement(new StreamRecord<>(input));
            ((ActionExecutionOperator<Long, Object>) harness.getOperator())
                    .waitInFlightEventsFinished();
            return ((List<StreamRecord<Object>>) harness.getRecordOutput())
                    .stream().map(StreamRecord::getValue).collect(Collectors.toList());
        }
    }
}

File: python/flink_agents/e2e_tests/e2e_tests_integration/react_subagent_test.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################
"""Offline ReAct sub-agent jobs exercising Python actions through the JVM operator."""

import sys
from pathlib import Path
from typing import Any, Sequence

import pytest
from pydantic import BaseModel
from pyflink.common import Encoder
from pyflink.common.typeinfo import Types
from pyflink.datastream import KeySelector, StreamExecutionEnvironment
from pyflink.datastream.connectors.file_system import StreamingFileSink

from flink_agents.api.agents.agent import Agent
from flink_agents.api.agents.react_agent import ReActAgent
from flink_agents.api.agents.types import OutputSchema
from flink_agents.api.chat_message import ChatMessage, MessageRole
from flink_agents.api.chat_models.chat_model import (
    BaseChatModelConnection,
    BaseChatModelSetup,
)
from flink_agents.api.decorators import action
from flink_agents.api.events.event import Event, InputEvent, OutputEvent
from flink_agents.api.execution_environment import AgentsExecutionEnvironment
from flink_agents.api.resource import ResourceDescriptor, ResourceType
from flink_agents.api.runner_context import RunnerContext
from flink_agents.api.tools.tool import Tool


class InputKeySelector(KeySelector):
    """Keep the single test record under a stable key."""

    def get_key(self, value: Any) -> Any:
        """Return the input as its key."""
        return value


class Answer(BaseModel):
    """Structured child result."""

    answer: str


class ScriptedSetup(BaseChatModelSetup):
    """Pass the model name to a deterministic connection."""

    @property
    def model_kwargs(self) -> dict[str, Any]:
        """Select the parent or child script."""
        return {"model": self.model}


class ScriptedConnection(BaseChatModelConnection):
    """Drive one delegation and one child tool call, without a model service."""

    def chat(
        self,
        messages: Sequence[ChatMessage],
        tools: list[Tool] | None = None,
        output_schema: OutputSchema | None = None,
        **kwargs: Any,
    ) -> ChatMessage:
        """Validate scope and return the next scripted response."""
        model = kwargs["model"]
        last = messages[-1]
        if last.role == MessageRole.TOOL:
            expected = "child evidence" if model.startswith("child") else "child answer"
            if expected not in last.content:
                msg = f"Wrong scoped result: {last.content}"
                raise ValueError(msg)
            if model == "parent_json" and last.content != '[{"answer":"child answer"}]':
                msg = f"Structured child output lost its JSON shape: {last.content}"
                raise ValueError(msg)
            content = (
                '{"answer":"child answer"}'
                if model == "child_json"
                else f"{model} answer"
            )
            return ChatMessage(role=MessageRole.ASSISTANT, content=content)
        if model.startswith("child"):
            if (
                messages[0].content != "Literal {prompt} instructions"
                or last.content != "investigate"
            ):
                msg = "Child instructions or input were not preserved"
                raise ValueError(msg)
            name, arguments = "evidence", {}
        else:
            name, arguments = "_subagent_researcher", {"prompt": "investigate"}
        if [tool.name for tool in tools] != [name]:
            msg = "Callable metadata was not resolved in the current scope"
            raise ValueError(msg)
        return ChatMessage(
            role=MessageRole.ASSISTANT,
            tool_calls=[
                {
                    "id": "call-1",
                    "type": "function",
                    "function": {"name": name, "arguments": arguments},
                }
            ],
        )


class ChildTools:
    """Child-local tools."""

    @staticmethod
    def evidence() -> str:
        """Read evidence owned by the child."""
        return "child evidence"


class ParentTools:
    """Parent-local tools with colliding names."""

    @staticmethod
    def evidence() -> str:
        """Read evidence owned by the parent."""
        return "parent evidence"


class ExplicitParent(Agent):
    """Invoke the child from an async action."""

    @action(InputEvent.EVENT_TYPE)
    @staticmethod
    async def invoke(event: Event, ctx: RunnerContext) -> None:
        """Emit the full list of child outputs, or its failure reason."""
        child = ctx.get_resource("researcher", ResourceType.AGENT)
        prompt = (
            42 if InputEvent.from_event(event).input == "invalid" else "investigate"
        )
        result = await (await child.submit(ctx, {"prompt": prompt}))
        ctx.send_event(
            OutputEvent(
                output=result.result
                if result.success
                else f"failed:{result.error_message}"
            )
        )


def parent_agent(model_driven: bool, structured: bool = False) -> Agent:
    """Build a parent and a child with colliding model/tool resource names."""
    child = ReActAgent.for_subagent(
        chat_model=ResourceDescriptor(
            clazz=f"{ScriptedSetup.__module__}.{ScriptedSetup.__name__}",
            connection="connection",
            model="child_json" if structured else "child",
            tools=["evidence"],
        ),
        description="Research a task",
        instructions="Literal {prompt} instructions",
        output_schema=Answer if structured else None,
    )
    child.add_resource(
        "evidence", ResourceType.TOOL, Tool.from_callable(ChildTools.evidence)
    )
    parent = (
        ReActAgent(
            chat_model=ResourceDescriptor(
                clazz=f"{ScriptedSetup.__module__}.{ScriptedSetup.__name__}",
                connection="connection",
                model="parent_json" if structured else "parent",
                subagents=["researcher"],
            )
        )
        if model_driven
        else ExplicitParent()
    )
    parent.add_resource(
        "connection",
        ResourceType.CHAT_MODEL_CONNECTION,
        ResourceDescriptor(
            clazz=f"{ScriptedConnection.__module__}.{ScriptedConnection.__name__}"
        ),
    )
    parent.add_resource(
        "evidence", ResourceType.TOOL, Tool.from_callable(ParentTools.evidence)
    )
    parent.add_resource("researcher", ResourceType.AGENT, child)
    return parent


@pytest.mark.parametrize(
    ("model_driven", "structured", "failure"),
    [
        (False, False, False),
        (True, False, False),
        (True, True, False),
        (False, False, True),
    ],
)
def test_react_subagent_tool_loop(
    tmp_path: Path, model_driven: bool, structured: bool, failure: bool
) -> None:
    """Both invocation paths run the child loop and resume the parent."""
    env = StreamExecutionEnvironment.get_execution_environment()
    env.set_parallelism(1)
    env.set_python_executable(sys.executable)
    agents = AgentsExecutionEnvironment.get_execution_environment(env=env)
    output = (
        agents.from_datastream(
            input=env.from_collection(["invalid" if failure else "hello"]),
            key_selector=InputKeySelector(),
        )
        .apply(parent_agent(model_driven, structured))
        .to_datastream()
    )
    destination = tmp_path / "results"
    output.map(str, Types.STRING()).add_sink(
        StreamingFileSink.for_row_format(
            base_path=str(destination),
            encoder=Encoder.simple_string_encoder(),
        ).build()
    )
    agents.execute()
    lines = [
        line
        for path in destination.rglob("*")
        if path.is_file()
        for line in path.read_text().splitlines()
    ]
    if failure:
        assert "\n".join(lines).startswith("failed:")
        return
    assert lines == [
        "parent_json answer"
        if structured
        else "parent answer"
        if model_driven
        else "['child answer']"
    ]
